float32 distance
int32 dir
